url43a0865bb0_kevinjdolan_darwin_tgz-pJ8-tui_DriverJ8/src/darwInvest/news/utility/Stemmer.java

```diff
@@ -0,0 +1,562 @@
+
+   Release 3
+
+   Considerably revised 4/9/00 in the light of many helpful suggestions
+   from Brian Goetz of Quiotix Corporation (brian@quiotix.com).
+
+   Release 4
+
+ */
+
+/**
+ * Stemmer, implementing the Porter Stemming Algorithm
+ *
+ * The Stemmer class transforms a word into its root form.  The input
+ * word can be provided a character at time (by calling add()), or at once
+ * by calling one of the various stem(something) methods.
+ * 
+ * @author Martin Porter
+ * @modified Kevin Dolan - for stylistic consistency with darwInvest
+ */
+public class Stemmer
+{  
+	private char[] b;
+	private int i,     /* offset into b */
+	i_end, /* offset to end of stemmed word */
+	j, k;
+	private static final int INC = 50;
+
+	/* unit of size whereby b is increased */
+	public Stemmer()
+	{  
+		b = new char[INC];
+		i = 0;
+		i_end = 0;
+	}
+
+	/**
+	 * Add a character to the word being stemmed.  When you are finished
+	 * adding characters, you can call stem(void) to stem the word.
+	 */
+	public void add(char ch)
+	{  
+		if (i == b.length) {  
+			char[] new_b = new char[i+INC];
+			for (int c = 0; c < i; c++) new_b[c] = b[c];
+			b = new_b;
+		}
+		b[i++] = ch;
+	}
+	
+	
+	/** Adds wLen characters to the word being stemmed contained in a portion
+	 * of a char[] array. This is like repeated calls of add(char ch), but
+	 * faster.
+	 */
+	public void add(char[] w, int wLen)
+	{  
+		if (i+wLen >= b.length) {  
+			char[] new_b = new char[i+wLen+INC];
+			for (int c = 0; c < i; c++) 
+				new_b[c] = b[c];
+			b = new_b;
+		}
+		for (int c = 0; c < wLen; c++) 
+			b[i++] = w[c];
+	}
+	
+	/**
+	 * After a word has been stemmed, it can be retrieved by toString(),
+	 * or a reference to the internal buffer can be retrieved by getResultBuffer
+	 * and getResultLength (which is generally more efficient.)
+	 */
+	public String toString() { 
+		return new String(b,0,i_end); 
+	}
+	
+	/**
+	 * Returns the length of the word resulting from the stemming process.
+	 */
+	public int getResultLength() { 
+		return i_end; 
+	}
+	
+	/**
+	 * Returns a reference to a character buffer containing the results of
+	 * the stemming process.  You also need to consult getResultLength()
+	 * to determine the length of the result.
+	 */
+	public char[] getResultBuffer() { 
+		return b; 
+	}
+	
+	/* cons(i) is true <=> b[i] is a consonant. */
+	private final boolean cons(int i)
+	{  
+		switch (b[i]) {  
+			case 'a': 
+			case 'e':
+			case 'i': 
+			case 'o': 
+			case 'u': 
+				return false;
+			case 'y': 
+				return (i==0) ? true : !cons(i-1);
+			default: return true;
+		}
+	}
+	
+	/* m() measures the number of consonant sequences between 0 and j. if c is
+	      a consonant sequence and v a vowel sequence, and <..> indicates arbitrary
+	      presence,
+	
+	         <c><v>       gives 0
+	         <c>vc<v>     gives 1
+	         <c>vcvc<v>   gives 2
+	         <c>vcvcvc<v> gives 3
+	         ....
+	 */
+	private final int m() {  
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.